lib/src/https_openssl.rs

```diff
@@ -1121,32 +1106,129 @@ pub struct Listener {
+
+            (fingerprint, certificate_and_key)
+        };
+
+        let certificate = X509::from_pem(&*certificate_and_key.certificate.contents)
+            .map_err(|err| ListenerError::PemParseError(err.to_string()))?;
+        let key = PKey::private_key_from_pem(certificate_and_key.key.as_bytes())
+            .map_err(|err| ListenerError::PemParseError(err.to_string()))?;
+        let chain = certificate_and_key
+            .chain
+            .iter()
+            .filter_map(|chain| X509::from_pem(&*chain.contents).ok())
+            .collect::<Vec<_>>();
+
+        let versions = if !certificate_and_key.versions.is_empty() {
+            &certificate_and_key.versions
+        } else {
+            &self.config.versions
+        };
+
+        let (context, _) = match Self::create_context(
+            versions,
+            &self.config.cipher_list,
+            &certificate,
+            &key,
+            &chain,
+            self.resolver.to_owned(),
+            self.contexts.to_owned(),
+        ) {
+            None => return Err(ListenerError::BuildOpenSslError),
+            Some(context) => context,
+        };
+
+        let mut contexts = self
+            .contexts
+            .lock()
+            .map_err(|err| ListenerError::LockError(err.to_string()))?;
+
+        contexts.insert(fingerprint.to_owned(), context);
+
+        Ok(fingerprint)
+    }
+
+    fn remove_certificate(
+        &mut self,
+        opts: &sozu_command_lib::proxy::RemoveCertificate,
+    ) -> Result<(), Self::Error> {
+        {
+            let mut resolver = self
+                .resolver
+                .lock()
+                .map_err(|err| ListenerError::LockError(err.to_string()))?;
+
+            resolver
+                .remove_certificate(opts)
+                .map_err(|err| ListenerError::ResolverError(err))?;
+        }
+
+        let mut contexts = self
+            .contexts
+            .lock()
+            .map_err(|err| ListenerError::LockError(err.to_string()))?;
+
+        contexts.remove(&opts.fingerprint);
+
+        Ok(())
+    }
+}
+
 impl Listener {
     pub fn new(config: HttpsListener, token: Token) -> Listener {
-        let contexts: HashMap<CertificateFingerprint, TlsData> = HashMap::new();
-        let domains = TrieNode::root();
-        let fronts = Router::new();
-        let rc_ctx = Arc::new(Mutex::new(contexts));
-        let ref_ctx = rc_ctx.clone();
-        let rc_domains = Arc::new(Mutex::new(domains));
-        let ref_domains = rc_domains.clone();
+        let resolver = Arc::new(Mutex::new(GenericCertificateResolver::new()));
+        let contexts = Arc::new(Mutex::new(HashMap::new()));
 
         let (default_context, ssl_options): (SslContext, SslOptions) =
-            Self::create_default_context(&config, ref_ctx, ref_domains)
+            Self::create_default_context(&config, resolver.to_owned(), contexts.to_owned())
                 .expect("could not create default context");
```

Comment:
Let's be sure a panic doesn't stop the entire process, or that it is very unlikely.

Reply:
Yeah, but listener are created at boot, so it is not a big deal. Besides, it affect a lot of code and we could enhance this in another pull request.